### PR TITLE
Add front-end testing to the Karate project

### PR DESCRIPTION
**Pull request by GabrielFleischer**
Add testing for the two front-end subprojects.

I greatly encourage you to apply the process described in the README to make sure nothing is missing.

This PR also add common utility classes with the MessageBuffer that allows multiple messages to be received and to take only the ones we want to check with filters.

### CODE REVIEW

File: tests/karate/src/test/java/fe/create_lao.feature
Comment (BadrTad) [inline tests/karate/src/test/java/fe/create_lao.feature:15]: For the timeouts i 'll suggest configuring those in the Background 
`configure timeout 1000`

File: tests/karate/src/test/java/fe/net/mockbackend.feature
Comment (BadrTad) [inline tests/karate/src/test/java/fe/net/mockbackend.feature:9]: Same remark as above the port parameter can be placed in karte-config.js and later 
configured  with `* configure port 9005`

File: tests/karate/src/test/java/fe/net/mockbackend.feature
Comment (BadrTad) [inline tests/karate/src/test/java/fe/net/mockbackend.feature:19]: Those should typically be in the Background 😄

File: tests/karate/src/test/java/fe/net/mockbackend.feature
Comment (BadrTad) [inline tests/karate/src/test/java/fe/net/mockbackend.feature:20]: the configuration of after scenario/feature should be done in the Background don't you think so ?

File: tests/karate/src/test/java/logback-test.xml
Comment (BadrTad) [inline tests/karate/src/test/java/logback-test.xml:17]: Just wondering how long it took to figure that this was an issue 😂

File: tests/karate/src/test/java/fe/create_lao.feature
Comment (BadrTad) [inline tests/karate/src/test/java/fe/create_lao.feature:1]: Test looks clean

File: tests/karate/src/test/java/karate-config.js
Comment (BadrTad) [inline tests/karate/src/test/java/karate-config.js:17]: Good to know 👍

File: tests/karate/src/test/java/fe/utils/web.feature
Comment (BadrTad) [inline tests/karate/src/test/java/fe/utils/web.feature:31]: What if it does not work do we get stuck infinitely in the loop ?

**Review by BadrTad — COMMENTED** [review]
This is a huge work , well done. Testing other features now should be straightforward.

File: tests/karate/src/test/java/fe/utils/web.feature
Comment (GabrielFleischer) [reply tests/karate/src/test/java/fe/utils/web.feature:32]: Hmmm... it goes into an infinit loop...

I'll add a maximum number of tries :D

File: tests/karate/src/test/java/logback-test.xml
Comment (GabrielFleischer) [reply tests/karate/src/test/java/logback-test.xml:17]: I was just annoyed by the number of information in the console, so I changed that part. And afterward when resetting it to debug, I wrote it in lowercase, that wasn't an issue

File: tests/karate/src/test/java/fe/utils/web.feature
Comment (GabrielFleischer) [reply tests/karate/src/test/java/fe/utils/web.feature:32]: Done

File: tests/karate/src/test/java/common/net/MessageFilters.java
Comment (pierluca) [inline tests/karate/src/test/java/common/net/MessageFilters.java:27]: careful. This can fail if the JSON is formatted with a space between the `method` field and its value (it'd still be valid JSON and respecting to have `"method":  "aaa"`)
As such, consider parsing it as JSON.

File: tests/karate/src/test/java/common/net/MessageFilters.java
Comment (pierluca) [inline tests/karate/src/test/java/common/net/MessageFilters.java:19]: extra blank lines? why? :'(

File: tests/karate/src/test/java/common/net/MultiMsgWebSocketClient.java
Comment (pierluca) [inline tests/karate/src/test/java/common/net/MultiMsgWebSocketClient.java:13]: semantically, does it make sense to say that the client "is" a MessageBuffer as opposed to "has" a MessageBuffer? 
Do we really need the delegate functions as opposed to having a getter for a MessageBuffer ?

File: tests/karate/src/test/java/fe/net/ReplyMethods.java
Comment (pierluca) [inline tests/karate/src/test/java/fe/net/ReplyMethods.java:14]: again, don't rely on a fixed formatting of the JSON. It'll make the tests very brittle.

File: tests/karate/src/test/java/common/net/MessageQueue.java
Comment (pierluca) [inline tests/karate/src/test/java/common/net/MessageQueue.java:120]: I'm not sure I fully understand this implementation. it seems that we're always waiting all the way until timeout unless the data was retrieved immediately, is that the desired behavior?

File: tests/karate/src/test/java/common/net/MessageQueue.java
Comment (pierluca) [inline tests/karate/src/test/java/common/net/MessageQueue.java:111]: I'm sure we could find a better variable name ;)

**Review by pierluca — COMMENTED** [review]
I haven't ye thad time to go through the README, but here are a few remarks concerning the infrastructure itself.

File: tests/karate/src/test/java/common/net/MessageFilters.java
Comment (GabrielFleischer) [reply tests/karate/src/test/java/common/net/MessageFilters.java:27]: I wanted to avoid this for some reason, but I will have to do that

File: tests/karate/src/test/java/common/net/MessageFilters.java
Comment (GabrielFleischer) [reply tests/karate/src/test/java/common/net/MessageFilters.java:19]: :'(

File: tests/karate/src/test/java/common/net/MultiMsgWebSocketClient.java
Comment (GabrielFleischer) [reply tests/karate/src/test/java/common/net/MultiMsgWebSocketClient.java:13]: The Multi-Message Client stores incoming messages, so I don't see it being a Buffer as a problem.

Also, it is for the karate implementation.
I like the idea to have a test that looks like this :

```
Then def msg1 = client.take()
And def msg2 = client.take()
And def msg3 = client.take()
```

rather than 

```
Then def msg1 = client.getBuffer().take()
And def msg2 = client.getBuffer().take()
And def msg3 = client.getBuffer().take()
```


But, now that I think about it, we could also write 
```
Then def buffer = client.getBuffer()
And def msg1 = buffer.take()
And def msg2 = buffer.take()
And def msg3 = buffer.take()
```

File: tests/karate/src/test/java/common/net/MessageQueue.java
Comment (GabrielFleischer) [reply tests/karate/src/test/java/common/net/MessageQueue.java:120]: In java, wait is different from a sleep.

A wait can be notified by notify() or notifyAll() and it will unlock.
The notifyAll is called when a new message is received and added to the buffer.

I was not sure myself if this was working and thus I wrote the Tests (MessageQueueTest).

File: tests/karate/src/test/java/common/net/MessageQueue.java
Comment (GabrielFleischer) [reply tests/karate/src/test/java/common/net/MessageQueue.java:111]: Oh, a Typo.

I could rename it to msg and msgSupplier too

File: tests/karate/src/test/java/common/net/MessageQueue.java
Comment (pierluca) [reply tests/karate/src/test/java/common/net/MessageQueue.java]: Thanks. I learned something

File: tests/karate/src/test/java/fe/net/MockBackend.java
Comment (BadrTad) [inline tests/karate/src/test/java/fe/net/MockBackend.java:28]: Don't you think this should be initialized to null, and the tester should always set a replyProducer manual for each message he/she eants to handle.

File: tests/karate/src/test/java/fe/net/MockBackend.java
Comment (BadrTad) [inline tests/karate/src/test/java/fe/net/MockBackend.java:62]: So following above comment here the replyProducer would be basically set back to null

File: tests/karate/src/test/java/fe/net/ReplyMethods.java
Comment (BadrTad) [inline tests/karate/src/test/java/fe/net/ReplyMethods.java:19]: What about error templates ?

File: tests/karate/src/test/java/fe/create_lao.feature
Comment (BadrTad) [inline tests/karate/src/test/java/fe/create_lao.feature:5]: A comment here maybe explaining what is going about the karate tags system

File: tests/karate/src/test/java/common/net/MessageQueue.java
Comment (GabrielFleischer) [reply tests/karate/src/test/java/common/net/MessageQueue.java:10]: I don't think it is possible to dequeue the first element matching a filter with a Dequeu.

And as I want to provide a way to retrieve the first message with a certain type with taking the order into account, the List is needed

File: tests/karate/src/test/java/fe/net/MockBackend.java
Comment (GabrielFleischer) [reply tests/karate/src/test/java/fe/net/MockBackend.java:28]: I like thinking of the default behavior as the one that will be used in the most common case.

And for me the most common case is to always return a Result.

File: tests/karate/src/test/java/fe/net/MockBackend.java
Comment (GabrielFleischer) [reply tests/karate/src/test/java/fe/net/MockBackend.java:62]: I don't agree.

If we want to do a message by message response, we can use backend.send() directly inside the test.

It will be easier to understand.

File: tests/karate/src/test/java/fe/net/ReplyMethods.java
Comment (GabrielFleischer) [reply tests/karate/src/test/java/fe/net/ReplyMethods.java:19]: No test need it yet so I did not implement the error template yet. But when a new test needs it, we can create the template easily.

File: tests/karate/src/test/java/fe/create_lao.feature
Comment (GabrielFleischer) [reply tests/karate/src/test/java/fe/create_lao.feature:5]: If you think so, I can add a link and a short explaination yes.

**Review by pierluca — COMMENTED** [review]
I went through everything this time. I can't speak much for the Java code (not a language I'm an absolute expert in), and I'll let others review those aspects (@celinecamacho ?), but the overall design seems clean and I'd say very close to a mergeable state 👍 

I'm happy to approve once a few typos / documentation / comments are fixed, and consider other improvements in separate PRs.

File: tests/karate/src/test/java/fe/net/MockBackend.java
Comment (GabrielFleischer) [reply tests/karate/src/test/java/fe/net/MockBackend.java:62]: Ok

File: tests/karate/README.md
Comment (GabrielFleischer) [reply tests/karate/README.md]: Yes, I tried to clarify this part with an example

File: tests/karate/src/test/java/fe/create_lao.feature
Comment (GabrielFleischer) [reply tests/karate/src/test/java/fe/create_lao.feature:8]: This should be clearer

File: tests/karate/README.md
Comment (GabrielFleischer) [reply tests/karate/README.md]: Done

File: tests/karate/src/test/java/common/net/MessageFilters.java
Comment (GabrielFleischer) [reply tests/karate/src/test/java/common/net/MessageFilters.java]: Done

File: tests/karate/src/test/java/fe/net/MockBackend.java
Comment (GabrielFleischer) [reply tests/karate/src/test/java/fe/net/MockBackend.java:60]: I tried to explain it

**Review by pierluca — APPROVED** [review]
LGTM !

**Pull request merged**
